minetest/minetest
# Fix Android segfault when game exits before TouchScreenGUI is initalized

Minetest for Android crashes (segfaults) if a Lua error occurs while starting a local singleplayer or multiplayer game.

<details>
<summary>Logcat output</summary>
<pre>05-06 17:38:42.712 30853 30971 E Minetest: 2023-05-06 17:38:42: ERROR[Main]: ModError: Failed to load and run script from /storage/emulated/0/Android/data/net.minetest.minetest/files/Minetest/games/nodecore/mods/nc_api/init.lua:
05-06 17:38:42.712 30853 30971 E Minetest: 2023-05-06 17:38:42: ERROR[Main]: ...etest/files/Minetest/games/nodecore/mods/nc_api/init.lua:1: wat
05-06 17:38:42.712 30853 30971 E Minetest: 2023-05-06 17:38:42: ERROR[Main]: stack traceback:
05-06 17:38:42.712 30853 30971 E Minetest: 2023-05-06 17:38:42: ERROR[Main]:    [C]: in function 'error'
05-06 17:38:42.712 30853 30971 E Minetest: 2023-05-06 17:38:42: ERROR[Main]:    ...etest/files/Minetest/games/nodecore/mods/nc_api/init.lua:1: in main chunk
05-06 17:38:42.712 30853 30971 E Minetest: 2023-05-06 17:38:42: ERROR[Main]: Check debug.txt for details.
05-06 17:38:42.712 30853 30971 F libc    : Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x28 in tid 30971 (MinetestNativeT), pid 30853 (netest.minetest)
05-06 17:38:42.760 30996 30996 E DEBUG   : failed to read /proc/uptime: Permission denied
05-06 17:38:42.910 30996 30996 F DEBUG   : Process name is net.minetest.minetest, not key_process
05-06 17:38:42.910 30996 30996 F DEBUG   : keyProcess: 0
05-06 17:38:42.910 30996 30996 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
05-06 17:38:42.910 30996 30996 F DEBUG   : Build fingerprint: [censored]
05-06 17:38:42.910 30996 30996 F DEBUG   : Revision: '0'
05-06 17:38:42.910 30996 30996 F DEBUG   : ABI: 'arm64'
05-06 17:38:42.910 30996 30996 F DEBUG   : Timestamp: 2023-05-06 17:38:42.759786206+0200
05-06 17:38:42.910 30996 30996 F DEBUG   : Process uptime: 0s
05-06 17:38:42.910 30996 30996 F DEBUG   : Cmdline: net.minetest.minetest
05-06 17:38:42.910 30996 30996 F DEBUG   : pid: 30853, tid: 30971, name: MinetestNativeT  >>> net.minetest.minetest <<<
05-06 17:38:42.910 30996 30996 F DEBUG   : uid: 10272
05-06 17:38:42.910 30996 30996 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x28
05-06 17:38:42.910 30996 30996 F DEBUG   : Cause: null pointer dereference
05-06 17:38:42.910 30996 30996 F DEBUG   :     x0  0000000000000000  x1  0000000000000000  x2  0000007b13600000  x3  0000000000000009
05-06 17:38:42.910 30996 30996 F DEBUG   :     x4  0000000000000089  x5  0000000000000000  x6  b400007b1496de58  x7  6420726f66207478
05-06 17:38:42.910 30996 30996 F DEBUG   :     x8  0000000000000000  x9  bd9cc13ff9e9e369  x10 0000000000000089  x11 0000000000000001
05-06 17:38:42.910 30996 30996 F DEBUG   :     x12 0000007b87daff48  x13 0000000000000010  x14 c3a5c85c97cb3127  x15 0000000034155555
05-06 17:38:42.910 30996 30996 F DEBUG   :     x16 0000007b1a833120  x17 0000007b1a05ef5c  x18 0000007b223ce000  x19 b400007b137e0000
05-06 17:38:42.910 30996 30996 F DEBUG   :     x20 0000000000000000  x21 0000000000000270  x22 0000007b89bfc000  x23 b400007b83271e00
05-06 17:38:42.910 30996 30996 F DEBUG   :     x24 b400007b136a3000  x25 0000007b1a854f20  x26 0000007b89bfc000  x27 0000007b89bfb560
05-06 17:38:42.910 30996 30996 F DEBUG   :     x28 b400007b137e0000  x29 0000007b89bfb000
05-06 17:38:42.910 30996 30996 F DEBUG   :     lr  0000007b1a05f1e8  sp  0000007b89bfb000  pc  0000007b1a05efac  pst 0000000060001000
05-06 17:38:42.910 30996 30996 F DEBUG   : backtrace:
05-06 17:38:42.910 30996 30996 F DEBUG   :       #00 pc 000000000064dfac  /data/app/~~sxtWeupGoj2drBgFvhT2gg==/net.minetest.minetest-5dcazSSMBfhJ945QCaEw9g==/lib/arm64/libMinetest.so (TouchScreenGUI::Toggle(bool)+80) (BuildId: 552a610c2a78141855078da3b94959ebe873e925)
05-06 17:38:42.910 30996 30996 F DEBUG   :       #01 pc 000000000064e1e4  /data/app/~~sxtWeupGoj2drBgFvhT2gg==/net.minetest.minetest-5dcazSSMBfhJ945QCaEw9g==/lib/arm64/libMinetest.so (TouchScreenGUI::hide()+20) (BuildId: 552a610c2a78141855078da3b94959ebe873e925)
05-06 17:38:42.910 30996 30996 F DEBUG   :       #02 pc 000000000054bbd8  /data/app/~~sxtWeupGoj2drBgFvhT2gg==/net.minetest.minetest-5dcazSSMBfhJ945QCaEw9g==/lib/arm64/libMinetest.so (Game::shutdown()+64) (BuildId: 552a610c2a78141855078da3b94959ebe873e925)
05-06 17:38:42.910 30996 30996 F DEBUG   :       #03 pc 00000000005578c0  /data/app/~~sxtWeupGoj2drBgFvhT2gg==/net.minetest.minetest-5dcazSSMBfhJ945QCaEw9g==/lib/arm64/libMinetest.so (the_game(bool*, InputHandler*, RenderingEngine*, GameStartData const&, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >&, ChatBackend&, bool*)+132) (BuildId: 552a610c2a78141855078da3b94959ebe873e925)
05-06 17:38:42.910 30996 30996 F DEBUG   :       #04 pc 0000000000512948  /data/app/~~sxtWeupGoj2drBgFvhT2gg==/net.minetest.minetest-5dcazSSMBfhJ945QCaEw9g==/lib/arm64/libMinetest.so (ClientLauncher::run(GameStartData&, Settings const&)+3764) (BuildId: 552a610c2a78141855078da3b94959ebe873e925)
05-06 17:38:42.910 30996 30996 F DEBUG   :       #05 pc 00000000007beab4  /data/app/~~sxtWeupGoj2drBgFvhT2gg==/net.minetest.minetest-5dcazSSMBfhJ945QCaEw9g==/lib/arm64/libMinetest.so (main+8076) (BuildId: 552a610c2a78141855078da3b94959ebe873e925)
05-06 17:38:42.910 30996 30996 F DEBUG   :       #06 pc 00000000008018a4  /data/app/~~sxtWeupGoj2drBgFvhT2gg==/net.minetest.minetest-5dcazSSMBfhJ945QCaEw9g==/lib/arm64/libMinetest.so (android_main+120) (BuildId: 552a610c2a78141855078da3b94959ebe873e925)
05-06 17:38:42.910 30996 30996 F DEBUG   :       #07 pc 0000000000dbcaf8  /data/app/~~sxtWeupGoj2drBgFvhT2gg==/net.minetest.minetest-5dcazSSMBfhJ945QCaEw9g==/lib/arm64/libMinetest.so (BuildId: 552a610c2a78141855078da3b94959ebe873e925)
05-06 17:38:42.910 30996 30996 F DEBUG   :       #08 pc 00000000000deb44  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+264) (BuildId: bbbdeb7c87c74f1491f92c6e605095b0)
05-06 17:38:42.910 30996 30996 F DEBUG   :       #09 pc 000000000007b2fc  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+68) (BuildId: bbbdeb7c87c74f1491f92c6e605095b0)</pre>
</details>

This happens because Minetest tries to hide and de-initialize its TouchScreenGUI even though TouchScreenGUI has not been initialized because of the Lua error. This is most likely is the same issue as #11732.

This PR just checks whether the TouchScreenGUI has been initialized before hiding and de-initializing it.

Closes #11732

## To do

This PR is a Ready for Review.

## How to test

Verify that Minetest for Android crashes if a Lua error occurs while starting a local singleplayer or multiplayer game. Verify that this doesn't happen anymore with this PR.
